test/unit/logic/transaction_pool.js

```diff
@@ -110,27 +170,1073 @@ describe('transactionPool', () => {
+				processVerifyTransactionStub
+			);
+			transactionPool.queueTransaction.callsArgWith(
+				1,
+				'Failed to queue bundled transaction'
+			);
+			transactionPool.processBundled(() => {
+				expect(logger.debug.args[0][0]).to.deep.eql(
+					'Failed to queue bundled transaction: 123'
+				);
+				expect(logger.debug.args[0][1]).to.deep.eql(
+					'Failed to queue bundled transaction'
+				);
+				done();
+			});
+		});
+	});
+
+	describe('undoUnconfirmedList', () => {
+		let undoUnconfirmedList;
+		let lastError;
+		let lastIds;
+
+		before(done => {
+			undoUnconfirmedList = transactionPool.undoUnconfirmedList;
+			done();
+		});
+
+		describe('when unconfirmed lists', () => {
+			describe('is empty', () => {
+				const transactions = [];
+
+				before(done => {
+					transactionPool.getUnconfirmedTransactionList = function() {
+						return transactions;
+					};
+
+					undoUnconfirmedList((err, ids) => {
+						lastError = err;
+						lastIds = ids;
+						done();
+					});
+				});
+
+				it('should not return an error', () => {
+					return expect(lastError).to.not.exist;
+				});
+
+				it('should not log an error', () => {
+					return expect(logger.error.called).to.be.false;
+				});
+
+				it('should return empty ids array', () => {
+					expect(lastIds).to.be.an('array');
+					return expect(lastIds.length).to.equal(0);
+				});
+
+				describe('modules.transactions.undoUnconfirmed', () => {
+					it('should not be called', () => {
+						return expect(dummyUndoUnconfirmed.called).to.be.false;
+					});
+				});
+
+				after(resetStates);
+			});
+
+			describe('contains 1 transaction', () => {
+				describe('that is valid', () => {
+					const validTransaction = { id: 'validTx' };
+					const transactions = [validTransaction];
+
+					before(done => {
+						transactionPool.addUnconfirmedTransaction(validTransaction);
+						transactionPool.getUnconfirmedTransactionList = function() {
+							return transactions;
+						};
+
+						undoUnconfirmedList((err, ids) => {
+							lastError = err;
+							lastIds = ids;
+							done();
+						});
+					});
+
+					it('should not return an error', () => {
+						return expect(lastError).to.not.exist;
+					});
+
+					it('should not log an error', () => {
+						return expect(logger.error.called).to.be.false;
+					});
+
+					it('should return valid ids array', () => {
+						expect(lastIds).to.be.an('array');
+						return expect(lastIds).to.deep.equal(
+							_.map(transactions, tx => {
+								return tx.id;
+							})
+						);
+					});
+
+					describe('modules.transactions.undoUnconfirmed', () => {
+						it('should be called onece', () => {
+							return expect(dummyUndoUnconfirmed.calledOnce).to.be.true;
+						});
+
+						it('should be called with transaction as parameter', () => {
+							return expect(dummyUndoUnconfirmed.args[0][0]).to.deep.equal(
+								validTransaction
+							);
+						});
+					});
+
+					describe('lists', () => {
+						let index;
```

Comment:
unconfirmed.index, queued.index and multisignature.index are out of the scope of this unit test